File: templates/python/before/chrome_custom_driver.py
# coding:utf-8
import os as Os, sys as Sys, types as Types, re as Re, time as Time, json as Json
from datetime import datetime as Datetime
from PIL import Image
from selenium import webdriver as Webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait, Select

if Os.path.exists('context.py'):
    from context import Context
    context = Context()
else:
    context = type('lamdbaobject', (object,), {})()

is_import = hasattr(context, 'driver')

if not hasattr(context, 'driver') or not hasattr(context, 'wait'):
    chromeOptions = Webdriver.ChromeOptions()
    download_directory = Os.getcwd()+'/__DOWNLOADDIR__'
    if Os.name == 'nt':
        download_directory = download_directory.replace('/', '\\')
    chromeOptions.add_experimental_option("prefs", {"download.default_directory" : download_directory})
    context.driver = Webdriver.Chrome(chrome_options=chromeOptions)
    context.wait = WebDriverWait(context.driver, __WAITTIMEOUT__)

def save_screenshot(context, path):
    scroll_point = 0
    screenshot_index = 0
    window_height = context.execute_script("return window.innerHeight")
    window_width = context.execute_script("return window.innerWidth")
    page_height = context.execute_script("return document.documentElement.scrollHeight")
    while page_height > scroll_point:
        context.execute_script("scroll(0, {0});".format(scroll_point))
        Time.sleep(1)
        context.save_screenshot_old(u"{0}.{1}.png".format(path, screenshot_index))
        scroll_point += window_height
        screenshot_index += 1
    base_image = Image.new('RGB', (window_width, page_height), (255, 255, 255))
    images = map(lambda index: u"{0}.{1}.png".format(path, index), xrange(0, screenshot_index))
    scroll_point = 0
    for index, filename in enumerate(images):
        if (len(images) == index + 1): scroll_point = page_height - window_height
        pasteImage = Image.open(filename, 'r')
        base_image.paste(pasteImage, (0, scroll_point))
        scroll_point += window_height
        Os.remove(filename)
    base_image.save(path, 'PNG', quality=100, optimize=True)
    context.execute_script("scroll(0, 0);")
context.driver.save_screenshot_old = context.driver.save_screenshot
context.driver.save_screenshot = Types.MethodType(save_screenshot, context.driver)

_driver = context.driver
_wait = context.wait

File: templates/ruby/before/chrome_custom_driver.rb
require 'selenium-webdriver'
require_relative "#{Dir.getwd}/context" if File.exist?("#{Dir.getwd}/context.rb")
require 'ostruct'
require 'openssl'
require 'net/http'
require 'RMagick'
include Magick

context = (defined?(Context) != nil) ? Context.instance : OpenStruct.new

is_import = context.driver != nil

module ChromeDriver
  def save_screenshot(path)
    scroll_point = 0
    screenshot_index = 0
    window_height = self.execute_script("return window.innerHeight")
    window_width = self.execute_script("return window.innerWidth")
    page_height = self.execute_script("return document.documentElement.scrollHeight")
    while page_height > scroll_point
      self.execute_script("scroll(0, #{scroll_point});")
      sleep 1
      super("#{path}.#{screenshot_index}.png")
      scroll_point += window_height
      screenshot_index += 1
    end
    base_image = Magick::Image.new(window_width, page_height)
    images = (0..screenshot_index-1).map {|index| "#{path}.#{index}.png" }
    scroll_point = 0
    images.each_with_index{|filename, index|
      scroll_point = page_height - window_height if (images.count == index + 1)
      img = Magick::Image.from_blob(File.read(filename)).shift
      base_image.composite!(img, 0, scroll_point, Magick::OverCompositeOp)
      scroll_point += window_height
      File.unlink filename
    }
    base_image.write(path)
    self.execute_script("scroll(0, 0);")
  end
end

if context.driver == nil || context.wait == nil
  download_directory = File.expand_path('__DOWNLOADDIR__')
  download_directory.gsub!("/", "\\") if  Selenium::WebDriver::Platform.windows?
  context.driver = Selenium::WebDriver.for :chrome, :prefs => {
    :download => {
      :prompt_for_download => false,
      :default_directory => download_directory
    }
  }
  context.driver.extend ChromeDriver
  context.wait = Selenium::WebDriver::Wait.new(:timeout => __WAITTIMEOUT__)
end

_driver = context.driver
_wait = context.wait

